ffi/src/scan.rs

```diff
@@ -0,0 +1,392 @@
+    // Item = Box<dyn EngineData>, see above, Vec<bool> -> can become a KernelBoolSlice
+    data: Box<dyn Iterator<Item = DeltaResult<(Box<dyn EngineData>, Vec<bool>)>>>,
+
+    // Also keep a reference to the external client for its error allocator.
+    // Parquet and Json handlers don't hold any reference to the tokio reactor, so the iterator
+    // terminates early if the last table client goes out of scope.
+    engine_interface: Arc<dyn ExternEngineInterface>,
+}
+
+impl BoxHandle for KernelScanDataIterator {}
+
+impl Drop for KernelScanDataIterator {
+    fn drop(&mut self) {
+        debug!("dropping KernelScanDataIterator");
+    }
+}
+
+/// Get an iterator over the data needed to perform a scan. This will return a
+/// [`KernelScanDataIterator`] which can be passed to [`kernel_scan_data_next`] to get the actual
+/// data in the iterator.
+///
+/// # Safety
+///
+/// Engine is responsible for passing a valid [`ExternEngineInterfaceHandle`] and [`Scan`]
+#[no_mangle]
+pub unsafe extern "C" fn kernel_scan_data_init(
+    engine_interface: *const ExternEngineInterfaceHandle,
+    scan: *mut Scan,
+) -> ExternResult<*mut KernelScanDataIterator> {
+    kernel_scan_data_init_impl(engine_interface, scan).into_extern_result(engine_interface)
+}
+
+unsafe fn kernel_scan_data_init_impl(
+    engine_interface: *const ExternEngineInterfaceHandle,
+    scan: *mut Scan,
+) -> DeltaResult<*mut KernelScanDataIterator> {
+    let engine_interface = unsafe { ArcHandle::clone_as_arc(engine_interface) };
+    let boxed_scan = unsafe { Box::from_raw(scan) };
+    let scan = boxed_scan.kernel_scan;
+    let scan_data = scan.scan_data(engine_interface.table_client().as_ref())?;
+    let data = KernelScanDataIterator {
+        data: Box::new(scan_data),
+        engine_interface,
+    };
+    Ok(data.into_handle())
+}
+
+/// # Safety
+///
+/// The iterator must be valid (returned by [kernel_scan_data_init]) and not yet freed by
+/// [kernel_scan_data_free]. The visitor function pointer must be non-null.
+#[no_mangle]
+pub unsafe extern "C" fn kernel_scan_data_next(
+    data: &mut KernelScanDataIterator,
+    engine_context: *mut c_void,
+    engine_visitor: extern "C" fn(
+        engine_context: *mut c_void,
+        engine_data: *mut EngineDataHandle,
+        selection_vector: KernelBoolSlice,
+    ),
+) -> ExternResult<bool> {
+    kernel_scan_data_next_impl(data, engine_context, engine_visitor)
+        .into_extern_result(data.engine_interface.error_allocator())
+}
+fn kernel_scan_data_next_impl(
+    data: &mut KernelScanDataIterator,
+    engine_context: *mut c_void,
+    engine_visitor: extern "C" fn(
+        engine_context: *mut c_void,
+        engine_data: *mut EngineDataHandle,
+        selection_vector: KernelBoolSlice,
+    ),
+) -> DeltaResult<bool> {
+    if let Some((data, sel_vec)) = data.data.next().transpose()? {
+        let bool_slice: KernelBoolSlice = sel_vec.into();
```

Comment:
It would be, but we don't implement `From` for the slices atm. If we decide to implement `From` we should do it that way.

Reply:
We can safely `impl From<Vec<bool>> for KernelBoolSlice` because the slice takes ownership of the data. 

I hesitated to `impl From<&str> for KernelStringSlice` because it wraps borrowed data and the associated lifetime info won't survive FFI crossings, and `From` seemed to make it easier to do something unwise. I only implemented `Into KernelStringSlice for &str` because it seemed like the least bad way to pass a `&str` as an arg to an FFI function that expects a `KernelStringSlice`. Very open to discussion about better approaches tho.

Reply:
No I think that makes sense. I've updated to that